source/common/upstream/upstream_impl.cc

```diff
@@ -208,6 +208,9 @@ void HostSetImpl::updateHosts(HostVectorConstSharedPtr hosts,
         locality_scheduler_->add(effective_weight, locality_entries_.back());
       }
     }
+    if (locality_scheduler_->empty()) {
```

Comment:
I'm still a bit confused here; in the above loop, we don't add entries to `locality_schduler_` if they have zero weight. How do we get into the situation where `locality_scheduler_` is populated with only entries with zero weight?

Reply:
I think there's a scenario where `locality_scheduler_` is created, but we don't populate it at all. I don't know the mechanism but I suspect there's a path where `effectiveLocalityWeight` returns 0 for all hosts, and no entries get written. This might not get caught by the preconditions in that outer if clause.

Reply:
We should be able to do a lighter weight check for `locality_scheduler_.empty()` that doesn't require us to cycle through all hosts, just looking at an empty queue?

Reply:
What if the queue contains localities but they are expired? Checking to see if the queue is empty doesn't guarantee that .pick() won't return a nullptr, I think.

Reply:
`locality_scheduler_` should be brand new in this function, as should any entries, so there should be no change for something to expire. This does raise an interesting bug though; we should probably perform a `locality_scheduler_.reset()` above the `if` statement, to ensure it is empty unless we explicitly initialize it. `updateHosts` is a "here's the entire new state of the world" operation, we should rebuild everything.